Add module "musig" that implements MuSig2 multi-signatures (BIP 327)

jonasnick opened the pull request:
EDIT: based on #1518. Closes #1452. Most of the code is a copy from [libsecp256k1-zkp](https://github.com/BlockstreamResearch/secp256k1-zkp). The API added in this PR is identical with the exception of two modifications:

1. I removed the unused `scratch_space` argument from `secp256k1_musig_pubkey_agg`. This argument was intended to allow using `ecmult_multi` algorithms for key aggregation in the future. But at this point it's unclear whether the `scratch_space` object will remain in its current form (see #1302).
2. Support for adaptor signatures was removed and therefore the `adaptor` argument of `musig_nonce_process` was also removed.

In contrast to the module in libsecp256k1-zkp, the module is non-experimental. I slightly cleaned up parts of the module, adjusted the code to the new definition of the VERIFY_CHECK macro and applied some simplifications that were possible because the module is now in the upstream repo (`ge_from_bytes`, `ge_to_bytes`). You can follow the changes I made to the libsecp256k1-zkp module at https://github.com/jonasnick/secp256k1-zkp/commits/musig2-upstream/.

jonasnick committed:
group: add ge_to_bytes and ge_from_bytes

jonasnick committed:
Add module "musig" that implements MuSig2 multi-signatures (BIP 327)

jonasnick committed:
build: allow enabling the musig module in cmake

real-or-random commented on include/secp256k1_extrakeys.h:
```diff
@@ -240,6 +240,22 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_keypair_xonly_tweak_add
     const unsigned char *tweak32
 ) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
 
+/** Compare two public keys using lexicographic order of their compressed
+ *  serialization.
+ *
+ *  Returns: <0 if the first public key is less than the second
+ *           >0 if the first public key is greater than the second
+ *           0 if the two public keys are equal
+ *  Args: ctx:      a secp256k1 context object.
+ *  In:   pubkey1:  first public key to compare
+ *        pubkey2:  second public key to compare
+ */
+SECP256K1_API int secp256k1_pubkey_cmp(
+    const secp256k1_context *ctx,
+    const secp256k1_pubkey *pk1,
+    const secp256k1_pubkey *pk2
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
```
Remind me what the difference was to the function `secp256k1_ec_pubkey_cmp`. Is there no difference and we simply overlooked this? I suspect what happened is this:
 - When we had x-only keys as input to key agg, we had to add a comparison function for x-only keys (to extrakeys module)
 - Then we switched to compressed keys as input, and we changed the comparison function to also take compressed keys, not noticing that there's one already in `secp256k1.h`

jonasnick replied:
There is no difference. I removed the commit.

real-or-random commented on src/secp256k1.c:
```diff
@@ -247,8 +247,15 @@ static int secp256k1_pubkey_load(const secp256k1_context* ctx, secp256k1_ge* ge,
     } else {
```
I think we should simply get rid of this branch entirely, see https://github.com/bitcoin-core/secp256k1/issues/1352. I can work on a PR next week.

real-or-random replied:
See #1480

jonasnick replied:
Yes thanks! I considered doing this but didn't want to make this PR dependent on a separate issue.

jonasnick replied:
Got rid of this by rebasing.

jonasnick commented:
Rebased on top of master to get https://github.com/bitcoin-core/secp256k1/pull/1480 which allowed dropping a commit. Old state is preserved at https://github.com/jonasnick/secp256k1/tree/musig2-module-backup.

real-or-random commented on include/secp256k1_extrakeys.h:
```diff
@@ -240,6 +240,21 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_keypair_xonly_tweak_add
     const unsigned char *tweak32
 ) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
 
+/** Sort public keys using lexicographic order of their compressed
+ *  serialization.
+ *
+ *  Returns: 0 if the arguments are invalid. 1 otherwise.
+ *
+ *  Args:     ctx: pointer to a context object
+ *  In:   pubkeys: array of pointers to pubkeys to sort
+ *      n_pubkeys: number of elements in the pubkeys array
+ */
+SECP256K1_API int secp256k1_pubkey_sort(
+    const secp256k1_context *ctx,
+    const secp256k1_pubkey **pubkeys,
+    size_t n_pubkeys
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2);
+
```
I think this function should be in the main module because it works on `secp256k1_pubkey` objects, and the comparison function is also there. (Extrakeys would make sense for x-only, I guess.) I don't think code size is an issue, the heap sort implementation should be tiny.

jonasnick replied:
That's reasonable. Done.

real-or-random commented on src/modules/musig/keyagg.h:
```diff
@@ -0,0 +1,40 @@
+    int parity_acc;
+} secp256k1_keyagg_cache_internal;
+
+/* musig_ge_to_bytes_ext and musig_ge_from_bytes_ext are identical to ge_save and ge_load
+ * except that they allow saving and loading the point at infinity */
+static void secp256k1_musig_ge_to_bytes_ext(unsigned char *data, secp256k1_ge *ge);
+
+static void secp256k1_musig_ge_from_bytes_ext(secp256k1_ge *ge, const unsigned char *data);
```
I think these should be in `group.h`, even though they'll be used only by the musig module. Infinity can be helpful in other contexts, and conceptually it's a group function.

Anyway, the comment needs to be updated `ge_save` and `ge_load` have been renamed.

jonasnick replied:
Agree, done.

real-or-random reviewed (commented):
needs rebase :)

t-bast commented:
FWIW, we have JVM bindings on top of this branch in https://github.com/ACINQ/secp256k1-kmp/pull/93 and an implementation of swap-in-potentiam (musig2 key-path with alternative delayed script path) in https://github.com/ACINQ/bitcoin-kmp/pull/107 and everything is working fine, and the API is easy enough to use!

jonasnick commented:
Rebased.

Thanks @t-bast, that's good to hear.

siv2r commented:
Attaching a visualization for the API flow.

![musig2-api-flowchart excalidraw](https://github.com/bitcoin-core/secp256k1/assets/56887198/12bf5cb8-9cb6-4f10-80db-75ef551c1fe5)

real-or-random commented on examples/musig.c:
```diff
@@ -0,0 +1,214 @@
+        pubkeys_ptr[i] = &signers[i].pubkey;
+    }
+    printf("ok\n");
+    printf("Combining public keys...");
```
I think there's no catch-all recommendation. BIP327 says "The aggregate public key produced by KeyAgg (regardless of the type) depends on the order of the individual public keys. If the application does not have a canonical order of the signers, the individual public keys can be sorted with the KeySort algorithm to ensure that the aggregate public key is independent of the order of signers."

In other words: If in your application, the collection of pubkeys (or signers represented by them) is conceptually an (ordered) list, then don't bother with sorting. If in your application, the collection of pubkeys is conceptually an (unordered) set, i.e., the application doesn't want to care about the order of pubkeys, then sort to make sure the set has a canonical serialization. 

Perhaps we can explain this somewhere in more detail, either in the API docs or in the example.

jonasnick replied:
Added sorting and a comment.

real-or-random commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,518 @@
+ *  MuSig differs from regular Schnorr signing in that implementers _must_ take
+ *  special care to not reuse a nonce. This can be ensured by following these rules:
+ *
+ *  1. Each call to this function must have a UNIQUE session_id32 that must NOT BE
+ *     REUSED in subsequent calls to this function.
+ *     If you do not provide a seckey, session_id32 _must_ be UNIFORMLY RANDOM
+ *     AND KEPT SECRET (even from other signers). If you do provide a seckey,
+ *     session_id32 can instead be a counter (that must never repeat!). However,
+ *     it is recommended to always choose session_id32 uniformly at random.
```
We should probably rename this. Despite this big fat warning, `session_id` itself sounds like a public value. 

Brain dump:
 -  `session_nonce`: too easy to confuse with the sec nonce
 -  `session_rand`: but what if it's counter
 -  `session_seed`: okayish

But now that I think about this again, I believe that the confusion stems from the fact that we have these two modes: It's either random and secret, or a counter, but that's only okay if you provide a seckey. A single name that fits both scenarios is necessarily imprecise. 

I think a better approach is to provide two different functions, e.g., `secp256k1_musig_nonce_gen` and `secp256k1_musig_nonce_gen_with_counter`. Then we can have clear argument names, even very verbose ones like `session_secret_rand` and `nonrepeating_counter`. Moreover, we can enforce the presence of the seckey in the counter function.

jonasnick replied:
Agreed. I split the two nonce_gen functions.

jonasnick commented on src/modules/musig/session_impl.h:
Because apparently I had been to lazy so far :D. I added this optimization.

real-or-random commented on include/secp256k1_musig.h:
```diff
@@ -317,12 +318,10 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_musig_pubkey_xonly_twea
- *     AND KEPT SECRET (even from other signers). If you do provide a seckey,
- *     session_id32 can instead be a counter (that must never repeat!). However,
- *     it is recommended to always choose session_id32 uniformly at random.
+ *     If you do not provide a seckey, session_secrand32 _must_ be UNIFORMLY RANDOM
+ *     AND KEPT SECRET (even from other signers).
```
We should then drop the condition "If you do not provide a seckey,"

real-or-random replied:
We could also add a reference to the new function, maybe below the numbered list.

something like "If you don't have access to good randomness, but you have access to a non-repeating counter, then see ..."

jonasnick replied:
Added both suggestions.

real-or-random commented on include/secp256k1_musig.h:
```diff
@@ -358,14 +357,72 @@ SECP256K1_API int secp256k1_musig_nonce_gen(
+ *  1. The nonrepeating_cnt argument must be a counter value that never
+ *     repeats, i.e., you must never call `secp256k1_musig_nonce_gen_counter`
+ *     twice with the same seckey and nonrepeating_cnt value.
+ *  2. If you already know the seckey, message or aggregate public key
+ *     cache, they can be optionally provided to derive the nonce and increase
+ *     misuse-resistance. The extra_input32 argument can be used to provide
```
seckey is mandatory for this function anyway.

I think usually I would suggest moving this big explainer at the top of the module, so that it covers both functions and does not need to be repeated. But in this case of a big fat warning, it's probably better to keep it in the docstring, where it's harder to miss...

jonasnick replied:
dropped mention of seckey

real-or-random commented on include/secp256k1_musig.h:
```diff
@@ -358,14 +357,72 @@ SECP256K1_API int secp256k1_musig_nonce_gen(
+ *  MuSig differs from regular Schnorr signing in that implementers _must_ take
+ *  special care to not reuse a nonce. This can be ensured by following these rules:
+ *
+ *  1. The nonrepeating_cnt argument must be a counter value that never
+ *     repeats, i.e., you must never call `secp256k1_musig_nonce_gen_counter`
+ *     twice with the same seckey and nonrepeating_cnt value.
```
We could expand a bit to mention that this includes cases where the same seckey is used on multiple devices. (Any other precautions to mention?)

jonasnick replied:
done

josibake reviewed (commented):
I'm also using the heapsort commits in #1471. What do you think about splitting out the sort commits into their own PR? Also fine with cherry-picking for now, but figured I'd mention it since it might simplify both of our PRs.

jonasnick commented on src/modules/extrakeys/main_impl.h:
```diff
@@ -282,4 +283,38 @@ int secp256k1_keypair_xonly_tweak_add(const secp256k1_context* ctx, secp256k1_ke
+    #if defined(_MSC_VER) && (_MSC_VER < 1933)
+    #pragma warning(push)
+    #pragma warning(disable: 4090)
+    #endif
```
You mean into `secp256k1_hsort`? I'd guess the wrong warning is emitted when `secp256k1_hsort` is called and therefore it would be to late when the warning was disabled in `secp256k1_hsort`.

jonasnick commented:
@josibake 

>  What do you think about splitting out the sort commits into their own PR?

That's a good idea. In particular, if more fixups are needed for the sort commits. See #1518.

achow101 commented:
Rebase perhaps? It'd be nice to get this in soon.

jonasnick committed:
group: add ge_to_bytes_ext and ge_from_bytes_ext

theStack reviewed (commented):
Looked mostly at the API header and the example so far, left a few nits. I noticed that the generated test vector headers differs in a few lines when generated with the data from the bips repo (master branch), probably that's due to the recent bump from 1.0.0 to 1.0.1 (https://github.com/bitcoin/bips/pull/1591).

real-or-random commented on configure.ac:
```diff
@@ -432,6 +444,10 @@ if test x"$enable_module_ecdh" = x"yes"; then
   SECP_CONFIG_DEFINES="$SECP_CONFIG_DEFINES -DENABLE_MODULE_ECDH=1"
 fi
 
+if test x"$enable_module_ellswift" = x"yes"; then
+  SECP_CONFIG_DEFINES="$SECP_CONFIG_DEFINES -DENABLE_MODULE_ELLSWIFT=1"
+fi
+
```
I think that's the duplication was introduced as a mistake when rebasing after https://github.com/bitcoin-core/secp256k1/pull/1482

jonasnick replied:
Yes, fixed

jonasnick commented on CHANGELOG.md:
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+   - Header file `include/secp256k1_musig.h` which defines the new API.
+   - Document `doc/musig.md` for further notes on API usage.
+   - Usage example `examples/musig.c`.
+ - Added `secp256k1_ec_pubkey_sort` which sorts an array of public keys (see `include/secp256k1.h`).
```
fixed

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,214 @@
+    while (1) {
+        if (!fill_random(seckey, sizeof(seckey))) {
+            printf("Failed to generate randomness\n");
+            return 1;
```
oops, fixed

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,214 @@
+    if (!secp256k1_musig_pubkey_ec_tweak_add(ctx, NULL, cache, plain_tweak)) {
+        return 0;
+    }
+    /* Note that we did not provided an output_pk argument, because the
```
done

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,214 @@
+        return 1;
+    }
+    printf("ok\n");
+    printf("Signing message.........");
```
Makes sense. Added fflush after every printf without a newline.

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+ * - 1 byte the parity of the internal key (if tweaked, otherwise 0)
+ * - 32 byte tweak
+ */
+/* Requires that cache_i->pk is not infinity and cache_i->second_pk_x to be normalized. */
```
it doesn't anymore, fixed

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    }
+}
+
+/* Assumes both field elements x and second_pk_x are normalized. */
```
This was a bogus comment anyway. Fixed.

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    secp256k1_ge_set_gej(&pkp, &pkj);
+    secp256k1_fe_normalize_var(&pkp.y);
+    /* The resulting public key is infinity with negligible probability */
+    VERIFY_CHECK(!secp256k1_ge_is_infinity(&pkp));
```
I believe so. These conditions are effectively unreachable. If they do become reachable, then we have a serious bug that the tests hopefully notice. Moreover, we cannot exercise the VERIFY_CHECKed conditions with a test, so if we would add a branch we couldn't cover it.

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+    int i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK((seckey != NULL));
```
done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+        if (!secp256k1_musig_partial_sig_load(ctx, &term, partial_sigs[i])) {
+            return 0;
+        }
+        secp256k1_scalar_add(&session_i.s_part, &session_i.s_part, &term);
```
We've already done the g*tacc computation in tweak_add and e*g*tacc in nonce_process.

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,214 @@
+
+/** This file demonstrates how to use the MuSig module to create a
+ *  3-of-3 multisignature. Additionally, see the documentation in
+ *  include/secp256k1_musig.h and src/modules/musig/musig.md.
```
done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+    ARG_CHECK(sig != NULL);
+    memcpy(out32, &sig->data[4], 32);
+    return 1;
+}
```
Added a memcmp. I don't think parsing the scalar is necessary because if properly initialized (which we ensure by checking the magic), the scalar cannot overflow. `_musig_{pub,agg}nonce_serialize` call load because they are required to do more work.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *  earlier failures).
+ *
+ *  secp256k1_musig_pubkey_agg(..., keyagg_cache, pubkeys, ...)
+ *  secp256k1_musig_pubkey_get(..., agg_pk, keyagg_cache)
```
Yep, fixed

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *  failures) returns 1.
+ *
+ *  secp256k1_musig_pubkey_agg(..., agg_pk, keyagg_cache, pubkeys, ...)
+ *  secp256k1_musig_pubkey_xonly_tweak_add(..., output_pk, tweak32, keyagg_cache)
```
fixed

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+    unsigned char zero[7] = { 0 };
+    /* The spec requires length prefixes to be between 1 and 8 bytes
+     * (inclusive) */
+    VERIFY_CHECK(prefix_size <= 8);
```
Done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+
+    for (i = 0; i < 8; ++i) {
+        buf[7 - i] = (nonrepeating_cnt >> (i * 8)) & 0xFF;
+    }
```
nice, done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+        memcpy(rand, session_secrand, sizeof(rand));
+    }
+
+    /* Subtract one from `sizeof` to avoid hashing the implicit null byte */
```
Yes, removed the comment

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+    secp256k1_gej_set_ge(&aggnonce_ptj[0], &aggnonce_pt[0]);
+    secp256k1_gej_set_ge(&aggnonce_ptj[1], &aggnonce_pt[1]);
+
+    if (!secp256k1_musig_nonce_process_internal(&session_i.fin_nonce_parity, fin_nonce, &session_i.noncecoef, aggnonce_ptj, agg_pk32, msg32)) {
```
Cool, implemented your suggestion.

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+        if (secp256k1_fe_is_odd(&cache_i.pk.y)) {
+            secp256k1_scalar_negate(&e_tmp, &e_tmp);
+        }
+        secp256k1_scalar_add(&session_i.s_part, &session_i.s_part, &e_tmp);
```
yes, done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+        secp256k1_musig_partial_sign_clear(&sk, k);
+        return 0;
+    }
+    secp256k1_fe_normalize_var(&pk.y);
```
Removed both calls to fe_normalize

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,803 @@
+        for (i = 0; i < 32; i++) {
+            rand[i] ^= seckey32[i];
+        }
+    } else {
+        memcpy(rand, session_secrand, sizeof(rand));
+    }
```
Done

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    if (agg_pk != NULL) {
+        memset(agg_pk, 0, sizeof(*agg_pk));
+    }
+    ARG_CHECK(pubkeys != NULL);
```
The SECP256K1_ARG_NONNULL attribute relies on a GNU extension that can give a warning at compile time. A failing ARG_CHECK will call the illegal_callback at runtime. They complement each other.

jonasnick commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1104 @@
+            CHECK((*tweak_func[i])(CTX, &tmp_output_pk, &tmp_keyagg_cache, tweak) == 1);
+            /* Reset keyagg_cache */
+            tmp_keyagg_cache = keyagg_cache;
+            CHECK((*tweak_func[i])(CTX, &tmp_output_pk, &tmp_keyagg_cache, tweak) == 1);
```
Yes, removed

jonasnick commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1104 @@
+    CHECK(memcmp_and_randomize(secnonce[0].data, zeros132, sizeof(secnonce[0].data)) == 0);
+    CHECK(secp256k1_musig_nonce_gen_counter(CTX, &secnonce[0], &pubnonce[0], nonrepeating_cnt, sk[0], &pk[0], msg, &keyagg_cache, NULL) == 1);
+
+    /* Every in-argument except session_secrand, sk and pubkey can be NULL */
```
fixed

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+    const unsigned char *msg32,
+    const secp256k1_musig_keyagg_cache *keyagg_cache,
+    const unsigned char *extra_input32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(5) SECP256K1_ARG_NONNULL(6);
```
nonrepeating_cnt isn't a pointer, so it cannot be NULL.

jonasnick commented on src/modules/musig/vectors.h:
```diff
@@ -0,0 +1,346 @@
+    unsigned char msg[32];
+    int has_extra_in;
+    unsigned char extra_in[32];
+    unsigned char expected_secnonce[97];
```
Hm the test vector is 97 bytes though. I changed the test to check that last 33 bytes of expected_secnonce as well, to make sure it's really the public key as expected.

jonasnick commented:
Thanks a lot @stratospher @theStack for the review! I think I addressed all comments and I rebased on master.

jonasnick commented:
rebased again

real-or-random commented on src/modules/musig/vectors.h:
```diff
@@ -0,0 +1,346 @@
+/**
+ * Automatically generated by ./contrib/musig2-vectors.py.
```
I think this was lost in some rebase. 

```suggestion
 * Automatically generated by tools/test_vectors_musig2_generate_.py.
```

And just remove `contrib/musig2-vectors.py`, the PR currently adds the (identical) script to both paths.

jonasnick replied:
Not a rebase issue, just confusion: I had thought that I had forgotten to add the python script because it's called differently in -zkp.

jonasnick replied:
fixed

real-or-random commented on doc/musig.md:
```diff
@@ -0,0 +1,51 @@
+The following sections contain additional notes on the API of the musig module (`include/secp256k1_musig.h`).
+A usage example can be found in `examples/musig.c`.
+
+# API misuse
```
nit:

```suggestion
## API misuse
```

(also for all other headings below)

`#` is the same level as the heading for the entire document, see the rendered file.

jonasnick replied:
fixed

real-or-random commented on examples/musig.c:
```diff
@@ -0,0 +1,218 @@
+/*************************************************************************
+ * Written in 2018 by Jonas Nick                                         *
```
(Okay, I'm all in favor of not doing updating years and stuff, but this is just too wrong. :D) 
```suggestion
 * Written by Jonas Nick                                                 *
```

jonasnick replied:
Removed name and year from all headers

real-or-random commented:
Just left a comment in the silent payments PR about checking test vectors, which is relevant here too: https://github.com/bitcoin-core/secp256k1/pull/1519#discussion_r1673629197

jonasnick commented:
> Just left a comment in the silent payments PR about checking test vectors, which is relevant here too: [#1519 (comment)](https://github.com/bitcoin-core/secp256k1/pull/1519#discussion_r1673629197)

This would only work if we included the json test vectors from the BIP.

real-or-random commented:
> This would only work if we included the json test vectors from the BIP.

Okay yeah, I think this has a non-zero benefit because then CI can test the py script and the repo is self-contained, but it's not clear if it's worth the hassle, in particular given that the PR is already in a pretty mature state.

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+ * - 32 byte tweak
+ */
+/* Requires that cache_i->pk is not infinity. */
+static void secp256k1_keyagg_cache_save(secp256k1_musig_keyagg_cache *cache, secp256k1_keyagg_cache_internal *cache_i) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Any reason `*cache_i` isn't const? You'd need to also the in param for both `_ge_to_bytes` and `_ge_to_bytes_ext` to be const, but seems like they should be unless I'm missing something.

josibake replied:
FWIW, I changed these functions to `const` and was able to compile and run the tests.

jonasnick replied:
Thanks, fixed.

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    ptr += 64;
+    memcpy(ptr, cache_i->pk_hash, 32);
+    ptr += 32;
+    *ptr = cache_i->parity_acc;
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Note to self/other reviewers: this initially tripped me up a bit as I expected something like `memcpy(ptr, &cache_i->parity_acc, 1);` but after some googling, apparently this is the more idiomatic/performant way of doing this!

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+}
+
+/* Computes pk_hash = tagged_hash(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_pk_hash(const secp256k1_context *ctx, unsigned char *pk_hash, const secp256k1_pubkey * const* pk, size_t np) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

nit: slightly more readable if `pk` is plural:

```suggestion
static int secp256k1_musig_compute_pk_hash(const secp256k1_context *ctx, unsigned char *pk_hash, const secp256k1_pubkey * const* pks, size_t n_pks) {
```

jonasnick replied:
done

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+
+    secp256k1_musig_keyagglist_sha256(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[33];
+        size_t ser_len = sizeof(ser);
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Could move this outside the loop to make this slightly more performant and, imo, more readable. Not sure the performance matters but could make a difference if computing `pk_hash` for a large number of keys?

jonasnick replied:
I think that would be less readable than declaring the variables only in the block where they are needed. The compiler will almost certainly optimize the current version to not cause any overhead.

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+
+    if (!secp256k1_ge_is_infinity(second_pk)
+          && secp256k1_fe_equal(&pk->x, &second_pk->x)
+          && secp256k1_fe_is_odd(&pk->y) == secp256k1_fe_is_odd(&second_pk->y)) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Not sure how important this check is , but I noticed this branch is not covered while checking `gcov` results. IIUC, the test case to add would be for pk and second_pk have the same x value  but even and odd y values.

jonasnick replied:
Good catch! We want everything covered by gcov. I replaced this with `ge_equals_ge_var` which was added after I wrote this code.

josibake reviewed (commented):
Working my way through `src/modules/musig/keyagg_impl.h`, left some suggestions but nothing blocking so feel free to ignore

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    secp256k1_scalar_get_b32(ptr, &cache_i->tweak);
+}
+
+static int secp256k1_keyagg_cache_load(const secp256k1_context* ctx, secp256k1_keyagg_cache_internal *cache_i, const secp256k1_musig_keyagg_cache *cache) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Note to self/reviewers: wasn't obvious to me how ctx was being used at first but it's being used in the `ARG_CHECK` macro.

josibake replied:
On second thought, seems a bit strange to be using `ARG_CHECK` inside an internal function (i.e., not exposed in the musig module API). Why not do an assert or a return here, which would let you simplify the function by removing `ctx` from the function signature?

josibake commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+    }
+}
+
+/* Outputs the point at infinity if the given byte array is all zero, otherwise
+ * attempts to parse compressed point serialization. */
+static int secp256k1_musig_ge_parse_ext(secp256k1_ge* ge, const unsigned char *in33) {
+    unsigned char zeros[33] = { 0 };
+
+    if (secp256k1_memcmp_var(in33, zeros, sizeof(zeros)) == 0) {
+        secp256k1_ge_set_infinity(ge);
+        return 1;
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Feels like I'm missing some context here. Generally, I'd expect setting the group element to the point at infinity to be a failure (i.e., expect the function to `return 0;`). But here it seems to be perfectly acceptable? My guess is that this failure mode is handled later on in a different function? It might be worth add some context in the comment for why this function is returning true, despite returning the point at infinity in the output param

jonasnick replied:
The point of the `serialize_ext` and `parse_ext` functions is to be able to serialize and parse the point at infinity. So this is not a failure case. This is noted in the doc of the function:

```
/* Outputs the point at infinity if the given byte array is all zero, otherwise
 * attempts to parse compressed point serialization. */
```

josibake commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+
+static const unsigned char secp256k1_musig_secnonce_magic[4] = { 0x22, 0x0e, 0xdc, 0xf1 };
+
+static void secp256k1_musig_secnonce_save(secp256k1_musig_secnonce *secnonce, const secp256k1_scalar *k, secp256k1_ge *pk) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

```suggestion
static void secp256k1_musig_secnonce_save(secp256k1_musig_secnonce *secnonce, const secp256k1_scalar *k, const secp256k1_ge *pk) {
```

jonasnick replied:
fixed (also in a few other places in session_impl.h)

josibake commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+    secp256k1_ge_to_bytes(&secnonce->data[68], pk);
+}
+
+static int secp256k1_musig_secnonce_load(const secp256k1_context* ctx, secp256k1_scalar *k, secp256k1_ge *pk, secp256k1_musig_secnonce *secnonce) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

```suggestion
static int secp256k1_musig_secnonce_load(const secp256k1_context* ctx, secp256k1_scalar *k, secp256k1_ge *pk, const secp256k1_musig_secnonce *secnonce) {
```

jonasnick replied:
fixed

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    if (output_pubkey != NULL) {
+        memset(output_pubkey, 0, sizeof(*output_pubkey));
+    }
+    ARG_CHECK(keyagg_cache != NULL);
+    ARG_CHECK(tweak32 != NULL);
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Generally speaking, seems like the `ARG_CHECKS` should be happening in the public facing functions, not internal functions? i.e., doing a runtime check inside a function an external caller doesn't have access to seems a bit strange to me.

jonasnick replied:
At the cost of code duplication?

josibake replied:
In this case, I think it helps with the clarity to have the arg checking happen in the public function _before_ calling the internal function (even at the cost of code duplication). 

That being said, this is really just a style nit so feel free to resolve this discussion if you prefer the (current) more concise version.

jonasnick replied:
I tried this suggestion and I prefer the more concise version.

josibake commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    return secp256k1_musig_pubkey_tweak_add_internal(ctx, output_pubkey, keyagg_cache, tweak32, 0);
+}
+
+int secp256k1_musig_pubkey_xonly_tweak_add(const secp256k1_context* ctx, secp256k1_pubkey *output_pubkey, secp256k1_musig_keyagg_cache *keyagg_cache, const unsigned char *tweak32) {
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

per my comment above, this is where I'd expect the `ARG_CHECKS` to happen, vs inside the internal function

josibake commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+
+static int secp256k1_musig_secnonce_load(const secp256k1_context* ctx, secp256k1_scalar *k, secp256k1_ge *pk, secp256k1_musig_secnonce *secnonce) {
+    int is_zero;
+    ARG_CHECK(secp256k1_memcmp_var(&secnonce->data[0], secp256k1_musig_secnonce_magic, 4) == 0);
```
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

This function could be simplified to return `void` if the `ARG_CHECK` were moved outside the function (seems like we could safely check these conditions before calling `_load`?)

josibake reviewed (commented):
Don't want to repeat the same comment in multiple places so I'll try to summarise here:

It seems there are quite a few internal functions where `ctx` is in the signature only for `ARG_CHECK` and/or functions that could be `void` return an error for `ARG_CHECK`. I'll revisit the public API of the module and take a look at the example to make sure I'm not missing something, but couldn't the magic bytes be checked before calling these internal functions and other `ARG_CHECKS` be moved to the public facing functions? The reason I think this is better is it simplifies the internal functions quite a bit and it also feels strange to me to be checking for illegal argument errors this deep in the internals of the module. It seems to me we should only get an illegal argument from a public facing function. Checking for illegal arguments inside internal functions feels indicates to me we aren't validating inputs earlier on in the public facing functions.

Again, I could be missing something here, so more just throwing this out for discussion.

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,217 @@
+    }
+    /* Note that we did not provide an output_pk argument, because the
+     * resulting pk is also saved in the cache and so if one is just interested
+     * in signing the output_pk argument is unnecessary. On the other hand, if
```
Fixed.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ *  secp256k1_musig_pubkey_ec_tweak_add(..., output_pk, tweak32, keyagg_cache)
+ *  secp256k1_ec_pubkey_serialize(..., buf, ..., output_pk, ...)
+ *  secp256k1_ec_pubkey_tweak_add(..., agg_pk, tweak32)
+ *  secp256k1_ec_pubkey_serialize(..., buf2, ..., agg_pk, ...)
```
Oops I missed that too, fixed.

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,217 @@
+    printf("Combining public keys...");
+    fflush(stdout);
+    /* If you just want to aggregate and not sign the cache can be NULL */
+    if (!secp256k1_musig_pubkey_agg(ctx, &agg_pk, &cache, pubkeys_ptr, N_SIGNERS)) {
```
Set agg_pk argument to NULL and improved comment.

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,217 @@
+    printf("ok\n");
+    secp256k1_context_destroy(ctx);
+    return 0;
+}
```
Thanks, secrets should be cleared everywhere now.

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+}
+
+/* Compute KeyAgg coefficient which is constant 1 for the second pubkey and
+ * otherwise tagged_hash(pk_hash, x) where pk_hash is the hash of public keys.
```
fixed

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,293 @@
+    secp256k1_extrakeys_ge_even_y(&pkp);
+    if (agg_pk != NULL) {
+        secp256k1_xonly_pubkey_save(agg_pk, &pkp);
+    }
```
done

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ * Further notes on API usage can be found in doc/musig.md
+ *
+ * Since the first version of MuSig is essentially replaced by MuSig2, we use
+ * MuSig, musig and MuSig2 synonymously unless noted otherwise.
```
I improved this doc comment in a way that's consistent with the existing doc.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ *  signer who is online throughout the whole process and can keep this
+ *  structure in memory can use the provided API functions for a safe standard
+ *  workflow. See
+ *  https://blockstream.com/2019/02/18/musig-a-new-multisignature-standard/ for
```
Yeah, I re-read the post and I'm not sure what exactly we wanted to convey here. The link predates the BIP. I think just saying that copying the nonce can leak the secret key is clear enough. I removed the link.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ *  Out:   out32: pointer to a 32-byte array to store the serialized signature
+ *  In:      sig: pointer to the signature
+ */
+SECP256K1_API int secp256k1_musig_partial_sig_serialize(
```
Let's minimize the triggers where reasonable ;P Changed.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ *  This function is required if you want to _sign_ for a tweaked aggregate key.
+ *  On the other hand, if you are only computing a public key, but not intending
+ *  to create a signature for it, you can just use
+ *  `secp256k1_ec_pubkey_tweak_add`.
```
done

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ *  This function is required if you want to _sign_ for a tweaked aggregate key.
+ *  On the other hand, if you are only computing a public key, but not intending
+ *  to create a signature for it, you can just use
+ *  `secp256k1_xonly_pubkey_tweak_add`.
```
done

jonasnick commented:
Thanks @josibake @theStack for the review! I should have addressed all comments except for the "don't use ARG_CHECK in internal functions one":

> It seems there are quite a few internal functions where ctx is in the signature only for ARG_CHECK and/or functions that could be void return an error for ARG_CHECK.

I agree that internal functions should avoid ARG_CHECKs where possible. In the case of the magic bytes, I think it's best to check the bytes where they are used, i.e., in the `_load` functions. If we'd check the magic bytes in the calling code, the code would be more fragile. A similar `ARG_CHECK` happens in `secp256k1_pubkey_load` in secp256k1.c.

josibake commented:
> I think it's best to check the bytes where they are used, i.e., in the _load functions

Thinking about this more (and taking a second look at the code), this is a really good point. Just to convince myself, I started working on a patch that moves the `_load` functions out of the internal functions wherever possible and instead passes `cache_i` as a parameter to the internal functions. The idea here was you could reason a `cache_i` has already been validated, which would simplify the logic of the internal functions and in some cases allows us to remove `ctx` from the function signatures and make some of the internal functions void. However, this ended up making `cache_i` an In/Out param and overall did feel more brittle. 

Feel free to resolve my review comments re: `ARG_CHECK`.

josibake commented on examples/musig.c:
```diff
@@ -0,0 +1,214 @@
+    secp256k1_musig_partial_sig partial_sig;
+};
+
+ /* Number of public keys involved in creating the aggregate signature */
+#define N_SIGNERS 3
```
I replaced the stack allocations with heap allocations here: https://github.com/josibake/secp256k1/commit/35ac3b194aefc064e8cee94e93498be224cc0d69

and ran the example with 100,000 signers. Everything up until signing was very fast, whereas signing took 45 mins (ran on a relatively powerful machine). Worth mentioning the time is a somewhat useless number here and the goal was more to stress test. The sign function in the example does both rounds of communication in the same function and the partial signing is done sequentially, whereas in reality the partial signing would be done in parallel

jonasnick replied:
I was a quite confused at first that signing took 45 minutes on your machine, until I realized that all signers in the example effectively do the coordinator's job of aggregating nonces. I changed that to only aggregate nonces once. Entire example with 100,000 signers takes 12 seconds on my laptop now.

josibake replied:
Ah, nice!

josibake reviewed (commented):
The example is great and really helped me understand how the full two round protocol is supposed to work. Left some nits, but nothing blocking

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,235 @@
+/* Create a key pair, store it in signer_secrets->keypair and signer->pubkey */
+static int create_keypair(const secp256k1_context* ctx, struct signer_secrets *signer_secrets, struct signer *signer) {
+    unsigned char seckey[32];
+    while (1) {
+        if (!fill_random(seckey, sizeof(seckey))) {
+            printf("Failed to generate randomness\n");
+            return 0;
+        }
```
Let's keep it consistent for now with the other examples.

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,235 @@
+        if (!secp256k1_musig_nonce_process(ctx, &session, &agg_pubnonce, msg32, cache)) {
+            return 0;
+        }
+        /* partial_sign will clear the secnonce by setting it to 0. That's because
+         * you must _never_ reuse the secnonce (or use the same session_secrand to
+         * create a secnonce). If you do, you effectively reuse the nonce and
+         * leak the secret key. */
+        if (!secp256k1_musig_partial_sign(ctx, &signer[i].partial_sig, &signer_secrets[i].secnonce, &signer_secrets[i].keypair, cache, &session)) {
+            return 0;
+        }
```
Hm, then people will copy the assert into their code...

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,235 @@
+static int tweak(const secp256k1_context* ctx, secp256k1_xonly_pubkey *agg_pk, secp256k1_musig_keyagg_cache *cache) {
+    secp256k1_pubkey output_pk;
+    unsigned char plain_tweak[32] = "this could be a BIP32 tweak....";
+    unsigned char xonly_tweak[32] = "this could be a taproot tweak..";
```
Good point. I updated the include file, example & musig.md to clarify that tweak32 is supposed to be a hash. Also added two sentences to the include file that say that the user is responsible for providing a tweak32 that is compatible with MuSig2 (although it is not known whether a fully attacker-controlled `tweak32` breaks unforgeability).

theStack reviewed (commented):
Went through another review round, LGTM, only found some nits that I left below. Haven't looked closer at CI changes and ctime-tests yet.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,584 @@
+ *
+ *  After the call, sig will always be initialized. If parsing failed or the
+ *  encoded numbers are out of range, signature verification with it is
+ *  guaranteed to fail for every message and public key.
```
I removed that paragraph.

  What we can provide to the user is the guarantee that even if the return value of partial_sig_parse isn't checked, then partial_sig_verify will for sure fail. Even though that was already the case due to the magic, I added an extra memset to make this extra clear.
I don't see a reason to mention this to the user (they should just check return values...).

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+    return 1;
+}
+
+int secp256k1_musig_pubnonce_serialize(const secp256k1_context* ctx, unsigned char *out66, const secp256k1_musig_pubnonce* nonce) {
```
fixed

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,584 @@
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Takes the public nonces of all signers and computes a session that is
+ *  required for signing and verification of partial signatures.
```
Wow this sentence was older than nonce aggregation. Fixed.

theStack reviewed (approved):
Light code-review ACK 398051cba6052c77f801b0aa57f1b0f161b0b12d

Reviewed the implementation code thoroughly and verified that it matches the BIP327 specification, and that all best practices that I'm aware of are followed (in particular, cleaning of secret data). Spent only a relatively small time looking at test and CI code.

real-or-random commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,805 @@
+        acc |= session_secrand32[i];
+    }
+    ret &= !!acc;
+    memset(&acc, 0, sizeof(acc));
```
I guess dead store elimination happens at a later compilation stage that does not care whether the store came from an assignment or a memset, so they're equally likely to be optimized out by the compiler.

So unfortunately both variants are without effect. The advantage of an explicit memset is that it stands out / is grepable, if something wants to pick up #636. (Feel free to give it a shot :P)

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,805 @@
+}
+
+/* tagged_hash(aggnonce[0], aggnonce[1], agg_pk, msg) */
+static int secp256k1_musig_compute_noncehash(unsigned char *noncehash, secp256k1_ge *aggnonce, const unsigned char *agg_pk32, const unsigned char *msg) {
```
fixed

jonasnick commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1110 @@
+    CHECK_ILLEGAL(CTX, secp256k1_musig_nonce_process(CTX, &session, &aggnonce, msg, &invalid_keyagg_cache));
+    CHECK(secp256k1_musig_nonce_process(CTX, &session, &aggnonce, msg, &keyagg_cache) == 1);
+
+    CHECK(secp256k1_musig_nonce_process(CTX, &session, &aggnonce, msg, &keyagg_cache) == 1);
```
fixed

jonasnick commented on src/ctime_tests.c:
```diff
@@ -180,6 +185,58 @@ static void run_tests(secp256k1_context *ctx, unsigned char *key) {
+        CHECK(secp256k1_musig_nonce_agg(ctx, &aggnonce, pubnonce_ptr, 1));
+        /* Make sure that previous tests don't undefine msg. It's not used as a secret here. */
+        SECP256K1_CHECKMEM_DEFINE(msg, sizeof(msg));
+        CHECK(secp256k1_musig_nonce_process(ctx, &session, &aggnonce, msg, &cache) == 1);
```
The return value doesn't depend on secret data.

jonasnick commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1110 @@
+    test_sha256_eq(&sha, sha_tagged);
+}
+
+/* Checks that the initialized tagged hashes initialized have the expected
```
fixed

jonasnick commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1110 @@
+    {
+        char tag[11] = "KeyAgg list";
+        secp256k1_musig_keyagglist_sha256(&sha);
+        sha256_tag_test_internal(&sha, (unsigned char*)tag, sizeof(tag));
```
done

theStack reviewed (approved):
Light re-ACK 5dccc7b4578fb583df60290fc363220cbacb7a03

josibake commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+
+/* Saves two group elements into a pubnonce. Requires that none of the provided
+ * group elements is infinity. */
+static void secp256k1_musig_pubnonce_save(secp256k1_musig_pubnonce* nonce, const secp256k1_ge* ge) {
```
in https://github.com/bitcoin-core/secp256k1/pull/1479/commits/2288e8808883556dc27cea6a61f409c08f70b30d:

nit: took me a second to mentally parse this function since the variable name is singular (ge), but it really is two group elements. Not a big deal since this is an internal function, but I think it would be more clear if the variable name were something like `ges` or `two_ges`. There are several other places where an array of `secp256k1_ge` objects is referred to as just `ge`, would be nice to make them all plural if you end up taking this suggestion.

jonasnick replied:
Ok, I should have replaced all the singulars with plurals.

josibake commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+}
+
+/* tagged_hash(aggnonce[0], aggnonce[1], agg_pk, msg) */
+static void secp256k1_musig_compute_noncehash(unsigned char *noncehash, secp256k1_ge *aggnonce, const unsigned char *agg_pk32, const unsigned char *msg) {
```
in https://github.com/bitcoin-core/secp256k1/pull/1479/commits/2288e8808883556dc27cea6a61f409c08f70b30d:

nit: same feedback regarding singular vs plural:

```suggestion
static void secp256k1_musig_compute_noncehash(unsigned char *noncehash, secp256k1_ge *aggnonces, const unsigned char *agg_pk32, const unsigned char *msg) {
```

josibake reviewed (approved):
ACK https://github.com/bitcoin-core/secp256k1/commit/5dccc7b4578fb583df60290fc363220cbacb7a03

Thanks for all the hard work on this, @jonasnick; I realise a lot of my feedback was on style 😅 Verified to the best of my ability that this conforms to the BIP327 spec, and that the module is well-tested (mainly by running and checking the `gcov` results and manually inspecting the tests). Overall, I found the module to be very well organised and easy to follow. I left one last nit regarding singular vs plural arguments, but not a blocker.

Bigheem commented on CHANGELOG.md:
```diff
@@ -11,6 +11,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Added
  - Added usage example for an ElligatorSwift key exchange.
+ - New module `musig` implements the MuSig2 multisignature scheme according to the [BIP 327 specification](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki). See:
+   - Header file `include/secp256k1_musig.h` which defines the new API.
+   - Document `doc/musig.md` for further notes on API usage.
+   - Usage example `examples/musig.c`.
```
bip-0174.mediawiki

jonasnick replied:
Oh, done!

Kixunil commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ */
+typedef struct {
+    unsigned char data[36];
+} secp256k1_musig_partial_sig;
```
It's unclear to me if these structs have any alignment needs. On the surface it looks like they don't but that might make the performance worse on some architectures? If it's not the case then it could be still nice to explicitly say they are unaligned in the doc. If it's the case maybe instead align them and say it in the doc?

real-or-random replied:
Hm, I don't see how anyone could get the idea that any type in a public header would have stricter alignment than what C anyway requires (unless otherwise specified in the header, of course). And The implied alignment is 1 in this case.

As far as I remember, we "read" from these char arrays into our internal data structures (the performance penalty is probably negligible), and so alignment shouldn't matter for performance either.

Kixunil replied:
Someone coming from Rust and being paranoid about everything. :D Thanks for clarifying, though I think a documentation comment would still be nice.

Kixunil commented:
I have a question that's maybe more about MuSig itself not this PR but it may have implications that could affect the API. But firstly I want to apologize for the question not being well-researched upfront, maybe this was already addressed somewhere, I don't have that much time right now and I feel like I should speak up before this merges and cements the API.

From my understanding the reason we have this "toxic data" of making sure nonces are not reused is that the counterparties could send nonces that are computed as sum of their nonce and inverse of our nonce which would allow them to compute private key. However can't the same be achieved by simply proving that they know the discrete log? Which IIUC could be done by simply single-signing the message they're going to sign (or some HMAC of it to avoid cryptographic interactions) producing a 64-byte proof. And once the nonces are proven to be random they can be deterministically derived from the message, the set of public keys and the private key. So the tradeoff would be to make all signing messages 64-byte larger to eliminate the risk that the state leaks somehow and with it also the keys or that nonces get reused by accident because of HW failure or a bug.

For an application I intend to use MuSig for I could easily have a kilobyte proof and it would still be worth the tradeoff. But I don't see any API here providing this. Sorry again if I missed it or if there's some serious problem in my reasoning that's documented somewhere. If this can be refuted with a link, you can just paste the link.

Thanks for consideration and all the hard work that went into this!

real-or-random commented:
> From my understanding the reason we have this "toxic data" of making sure nonces are not reused is that the counterparties could send nonces that are computed as sum of their nonce and inverse of our nonce which would allow them to compute private key.

I think you're mixing up two different issues. What you describe doesn't lead to an attack. The attacker can choose nonces that cancel out a honest signer's nonces, but the only effect this will have is that the attacker won't be able to come up with a valid partial signature and thus the protocol will fail.

This cancellation attack you describe is a concern not with nonces, but with the individual public keys in the key aggregation. Indeed, cancelling a key of an honest signer would work if the aggregate public key was just the sum of the individual keys. (This is called a "rouge key attack" or literally "a key cancellation attack"). But MuSig2 eliminates this attack vector by using a key aggregation function that is not just the sum of the keys, but instead a sum with random coefficients.

But yes, letting everyone prove knowledge of the discrete logarithm of their public individual public key using a Schnorr signature is a different way to eliminate rogue key attacks. This is used in SpeedyMuSig: https://eprint.iacr.org/2021/1375.pdf The drawback of this method is that it doesn't suffice to send these additional Schnorr signatures around during the signing protocol. For this to be secure, the signatures need to be there already at the key aggregation stage, which is possibly performed by a different party. In MuSig2, anyone (i.e., not just the involved signers) can combine some keys A and B into an aggregate key and use it, e.g., send money to it, and it will be ensured that A and B can only spend it if they work together. In SpeedyMuSig, the party who performs the key aggregation needs to see the Schnorr signatures valid for A and B before they can securely send money to the aggregate key A+B. This complicates key management and is a potential footgun (people could send to A+B without checking the additional Schnorr sigs)

Kixunil commented:
Ah, OK I think I understand better now. Thanks for explaining! What I was really thinking was if all nonces are deterministic then for the same key and same message they produce the same signature so how could an attacker possibly extract additional information from that? What I was missing is there's no way for a signer to know if the pubnonce sent by another party was generated deterministically except for having some kind of ZKP that would be likely expensive to construct and verify.

Still it would be nice to not have to store the secnonce on a permanent storage so maybe a viable solution for cases when there's one "unreliable" signer (random user with a smartphone that could kill the app at any time) and other parties are "reliable" (run on server 24/7, can hold the secnonces in RAM) would be for the "unreliable" signer to collect all nonces first and compute his own as an HMAC(secret_key, message || aggregate_pubkey || all_nonces_except_signers) and be the last one who sends pubnonce. The other signers hold their secnonces in RAM in the meantime. Does that make sense?

jonasnick commented:
@Kixunil

> if all nonces are deterministic then for the same key and same message they produce the same signature so how could an attacker possibly extract additional information from that? What I was missing is there's no way for a signer to know if the pubnonce sent by another party was generated deterministically except for having some kind of ZKP that would be likely expensive to construct and verify.

Proving that the nonces were generated deterministically from the session parameters is the idea behind deterministic signing in [MuSig-DN](https://eprint.iacr.org/2020/1057) and [Exponent-VRFs and Their Applications](https://eprint.iacr.org/2024/397).

> Does that make sense?

It sounds like you're describing a variant of the ["Deterministic and Stateless Signing for a Single Signer"-mode](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#modifications-to-nonce-generation) of BIP MuSig. In this mode, one signer is able to derive the nonces deterministically from the session parameters and the pubnonces of all other signers.

jonasnick commented on src/modules/musig/keyagg.h:
```diff
@@ -0,0 +1,33 @@
+#include "../../../include/secp256k1.h"
+#include "../../../include/secp256k1_musig.h"
+
+#include "../../field.h"
```
done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+    /* TODO: use multiexp to compute -s*G + e*mu*pubkey + aggnonce[0] + b*aggnonce[1] */
+    if (!secp256k1_musig_pubnonce_load(ctx, nonce_pt, pubnonce)) {
+        return 0;
+    }
```
Yes, done.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,580 @@
+ *  and verification of partial signatures.
+ *
+ *  Returns: 0 if the arguments are invalid or if some signer sent invalid
+ *           pubnonces, 1 otherwise
```
Right, forgot about that. We can even make the type of nonce_process_internal  void.

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,801 @@
+    secp256k1_gej_set_ge(&aggnoncej[1], &aggnonce[1]);
+    /* fin_nonce = aggnonce[0] + b*aggnonce[1] */
+    secp256k1_scalar_set_b32(b, noncehash, NULL);
+    secp256k1_gej_set_infinity(&fin_nonce_ptj);
```
Yes, done. Also added your helper function.

Kixunil commented:
@jonasnick thanks for references!

> It sounds like you're describing a variant of the ["Deterministic and Stateless Signing for a Single Signer"-mode](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#modifications-to-nonce-generation) of BIP MuSig.

Yes, that's what I want! From what I can see it's not currently supported by the API. Is this planned or can it be implemented some other way?

Kixunil commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,579 @@
+SECP256K1_API int secp256k1_musig_pubnonce_parse(
+    const secp256k1_context *ctx,
+    secp256k1_musig_pubnonce *nonce,
+    const unsigned char *in66
```
Wouldn't it be better to declare this as `const unsigned char (*in66)[66]`? Aside from clarity it could help static analysis tools or perhaps bindings generators to generate correct bindings.

jonasnick replied:
That's an interesting point. Right now, we consistently use declarations of the form `const unsigned char *in66` consistently in the library. I prefer to have a standard across the whole API and would avoid just changing this in the musig module.

Kixunil replied:
That makes sense. Might be nice to change all of them at some point unless there's a problem I'm missing.

Kixunil replied:
For instance in Rust we have [`bindgen`](https://crates.io/crates/bindgen), a tool that generates bindings automatically. I've tried to pass it this header file:

```c
void foo(const unsigned char *arg);
void bar(const unsigned char arg[32]);
void baz(const unsigned char (*arg)[32]);
```

And it generated this code:

```rust
extern "C" {
    pub fn foo(arg: *const ::std::os::raw::c_uchar);
}
extern "C" {
    pub fn bar(arg: *const ::std::os::raw::c_uchar);
}
extern "C" {
    pub fn baz(arg: *const [::std::os::raw::c_uchar; 32usize]);
}
```

As you can see, the last one has a different type. And while it's the same from ABI perspective, there's a real difference between them in Rust since you can directly pass in `&array` without any casts because the types match making the code more obviously correct.

Currently, we aren't using `bidgen` in rust-secp256k1 but we might in the future and I'd be surprised if similar tools don't exist in other languages.

Also I'm not familiar with the current state of static analysis in C but I think a reasonable static analyzer should be able to see that statements like `out[33] = 42` or `baz(&twentyone_item_array)` are obviously wrong. `gcc` already complains about the these when compiling with `-O2 -W -Wall -Warray-bounds`.

Kixunil replied:
Oh, I found that `bindgen` has an option to generate `*const [c_uchar; 32]` instead of `*const c_uchar` for arguments defined as `const unsigned char arg[32]`, so maybe even that is already valuable? (IIUC not API-breaking) I think it has documentation value too.

real-or-random replied:
@Kixunil We discussed this in our IRC meeting today. Can you try if bindgen understands `arg[static 32]`? This is a special syntax in C99, see https://hamberg.no/erlend/posts/2013-02-18-static-array-indices.html

sipa reviewed (commented):
Mostly looked through the interface and documentation.

jonasnick commented on doc/musig.md:
```diff
@@ -0,0 +1,51 @@
+2. Call `secp256k1_musig_pubkey_agg` with the pubkeys of all participants.
+3. Optionally add a (Taproot) tweak with `secp256k1_musig_pubkey_xonly_tweak_add` and a plain tweak with `secp256k1_musig_pubkey_ec_tweak_add`.
+4. Generate a pair of secret and public nonce with `secp256k1_musig_nonce_gen` and send the public nonce to the other signers.
+5. Someone (not necessarily the signer) aggregates the public nonce with `secp256k1_musig_nonce_agg` and sends it to the signers.
```
fixed

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,579 @@
+ *     misuse-resistance. The extra_input32 argument can be used to provide
+ *     additional data that does not repeat in normal scenarios, such as the
+ *     current time.
+ *  3. Avoid copying (or serializing) the secnonce. This reduces the possibility
```
We don't provide the ability to serialize the secnonce right now, which prevents users from storing/loading from disk. Maybe worth a separate PR?

jonasnick replied:
If we provide the on store/load the secnonce from disk, we should also provide de/serialization functions because our API docs instruct users to not read the raw bytes. I had considered this out of scope for this initial PR because it seems dangerous. Especially since so far no one requested this feature, just "don't store to disk" seems to be a sensible answer.

Actually one developer asked me about the missing de/serialization functions for the secnonce but further discussion on their motivation for requesting this revealed that they planned to build a very insecure protocol because they had entirely misunderstood what nonce reuse actually is. In the end they agreed that they should not store the secnonce to disk and changed their implementation accordingly.

Also I had hoped that someone would eventually come up with a more clever API that can protect from misuse better than simple de/serialization functions (at least in some cases). But maybe that just doesn't exist.

jonasnick replied:
If someone really needs to do this right now, then I would probably recommend option 2. In contrast to option 1, if you reuse `session_secrand` but you're signing for a different message or aggregate public key, you're protected.

One downside of that method is that `session_secrand` has no "in-memory protection", in the sense that the `session_secrand` buffer is not cleared after nonce_generation. We do wipe the `secnonce` after signing. Maybe we should change nonce generation to overwrite the `session_secrand` buffer?

jonasnick replied:
Done. The secrand buffer is invalidated whenever nonce_gen returns 1. The alternative is to invalidate whenever a session_secrand argument is given, which is a tiny bit more tedious to test and document. Anyway, if nonce_gen returns 0, then the secnonce is invalid as well, so reusing the secrand buffer in that case is not an issue.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,579 @@
+ *  In:
+ *  session_secrand32: a 32-byte session_secrand32 as explained above. Must be unique to this
+ *                     call to secp256k1_musig_nonce_gen and must be uniformly random.
+ *             seckey: the 32-byte secret key that will later be used for signing, if
```
The musig_nonce_gen functions do not check if the seckey matches the given pubkey.

jonasnick replied:
If the wrong public key is provided, signing will fail. This works because the public key is appended to the secnonce output of the `nonce_gen` functions and `partial_sign` will compare this public key with the public key the signer is trying to sign for.

Providing the optional `seckey` argument is a defense-in-depth measure to strengthen nonce generation against a low entropy `session_secrand` argument. If provided, the `seckey` is simply hashed into the nonce and does not have an effect on whether the resulting signature is valid.

So, the idea behind the defense-in-depth measure is that if the `seckey` does correspond to the `pubkey` argument and has high entropy, but the `session_secrand` argument has low entropy (but does not repeat), the generated nonce is secure. On the other hand, if the `seckey` does not match the `pubkey`, nonce generation is only insecure if both the provided `seckey` and `secrand` have low entropy. Providing a wrong `seckey` cannot negatively affect nonce generation as long as `secrand` has enough entropy.

jonasnick replied:
> Perhaps we should note that a pubkey/seckey mismatch does not count as invalid, but that it will be caught later during signing.

I added a sentence, but it may be confusing. A pubkey/seckey mismatch in nonce generation may not be caught during signing, because the signer can still provide the correct seckey for signing.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,579 @@
+ *  This can be done by an untrusted party to reduce the communication
+ *  between signers. Instead of everyone sending nonces to everyone else, there
+ *  can be one party receiving all nonces, aggregating the nonces with this
+ *  function and then sending only the aggregate nonce back to the signers.
```
Yeah, done.

jonasnick commented on src/modules/musig/keyagg_impl.h:
```diff
@@ -0,0 +1,292 @@
+}
+
+/* Computes pks_hash = tagged_hash(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_pks_hash(const secp256k1_context *ctx, unsigned char *pks_hash, const secp256k1_pubkey * const* pks, size_t np) {
```
I think we only followed that style for externally facing functions, also in the Schnorr signature and extrakeys module. But it'd be possible to adopt that style and start with the musig module.

jonasnick commented on CHANGELOG.md:
```diff
@@ -7,10 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-## [0.5.1] - 2024-08-01
-
 #### Added
  - Added usage example for an ElligatorSwift key exchange.
```
:rage:  ... thanks ... should be fixed for real now

theStack reviewed (approved):
Code-review ACK fdc09608036822afc1cebbe0c5b56cebf8ba508d

Reviewed once again that the code matches the specification in BIP327 and also took a closer look on the tests. Regarding the interface discussion https://github.com/bitcoin-core/secp256k1/pull/1479#discussion_r1747513031, I wonder if it's acceptable for now to not support (and not even mention) the use-case where users can't store the `secnonce` in volatile memory by now, and leave that for a follow-up? If a suggested way to work around that is described in the API docs ("if you really need to do this..."), that could be seen by users as invitation to break the rules. On the other hand, if nothing is mentioned, users might break the rules anyway and do it in an even worse way, so not sure what is better. Anyways, happy to re-review on changes in that area (e.g. suggestions in https://github.com/bitcoin-core/secp256k1/pull/1479#discussion_r1761370580).

real-or-random commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,818 @@
+    return 1;
+}
+
+static int secp256k1_musig_sum_nonces(const secp256k1_context* ctx, secp256k1_gej *summed_nonces, const secp256k1_musig_pubnonce * const* pubnonces, size_t n_pubnonces) {
```
```suggestion
static int secp256k1_musig_sum_pubnonces(const secp256k1_context* ctx, secp256k1_gej *summed_pubnonces, const secp256k1_musig_pubnonce * const* pubnonces, size_t n_pubnonces) {
```
perhaps a slightly better name

jonasnick replied:
done

real-or-random reviewed (approved):
ACK fdc09608036822afc1cebbe0c5b56cebf8ba508d

I left a bunch of nits, mostly noticed when I looked at @sipa's comments in the include file, but none of these are blockers.

jonasnick committed:
util: add constant-time is_zero_array function

jonasnick commented on README.md:
```diff
@@ -21,6 +21,7 @@ Features:
 * Optional module for ECDH key exchange.
 * Optional module for Schnorr signatures according to [BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki).
 * Optional module for ElligatorSwift key exchange according to [BIP-324](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki).
+* Optional module for the MuSig2 multi-signature scheme according to [BIP-327](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki).
```
fixed

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *  341.
+ *
+ *  Callers are responsible for deriving `tweak32` in a way that does not reduce
+ *  the security of MuSig2 (for example, by following Taproot BIP 341).
```
done

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *  a hash as defined in BIP 32.
+ *
+ *  Callers are responsible for deriving `tweak32` in a way that does not reduce
+ *  the security of MuSig2 (for example, by following BIP 32).
```
done

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *  1. Each call to this function must have a UNIQUE session_secrand32 that must
+ *     NOT BE REUSED in subsequent calls to this function and must be KEPT
+ *     SECRET (even from other signers).
+ *  2. If you already know the seckey, message or aggregate public key
+ *     cache, they can be optionally provided to derive the nonce and increase
+ *     misuse-resistance. The extra_input32 argument can be used to provide
+ *     additional data that does not repeat in normal scenarios, such as the
+ *     current time.
```
done

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *  It is not required to call this function in regular MuSig sessions, because
+ *  if any partial signature does not verify, the final signature will not verify
+ *  either, so the problem will be caught. But this function allows determining
+ *  the specific party who produced an invalid signature.
```
done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,818 @@
+     * in addition to the magic. */
+    is_zero = secp256k1_scalar_is_zero(&k[0]) & secp256k1_scalar_is_zero(&k[1]);
+    secp256k1_declassify(ctx, &is_zero, sizeof(is_zero));
+    ARG_CHECK(!is_zero);
```
I implemented this by creating a `is_zero_array` function in util.h and using it from both places in the musig module. I think this is a bit nicer, because `is_zero_array` is reusable, but the change did not turn out to make the secnonce load function simpler. Maybe you meant something else?

jonasnick commented on src/group_impl.h:
```diff
@@ -963,4 +963,21 @@ static void secp256k1_ge_from_bytes(secp256k1_ge *r, const unsigned char *buf) {
+}
+
+static void secp256k1_ge_from_bytes_ext(secp256k1_ge *ge, const unsigned char *data) {
+    unsigned char zeros[64] = { 0 };
```
done

real-or-random replied:
nit: you did this, but in the "wrong" commit

jonasnick replied:
fixed

jonasnick commented on doc/musig.md:
```diff
@@ -0,0 +1,53 @@
+## API misuse
+
+The musig API is designed to be as misuse resistant as possible.
+However, the MuSig protocol has some additional failure modes (mainly due to interactivity) that do not appear in single-signing.
```
improved the wording

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *           invalid (only when the tweak is the negation of the corresponding
+ *           secret key). 1 otherwise.
+ *  Args:            ctx: pointer to a context object
+ *  Out:   output_pubkey: pointer to a public key to store the result. Will be set
```
It's called `output_pubkey` because that is the terminology we use for BIP 341 tweaking. People don't have that context, so I'm ok with changing it. But if we do, then how about `agg_pk` because that's what we're using currently for the aggregation and get_pubkey functions.

real-or-random replied:
> It's called `output_pubkey` because that is the terminology we use for BIP 341 tweaking. 

Oh, sure. Never mind, I mixed things up and thought this is simply the aggregate key. But it's the tweaked aggregate key, so yep, it's better to have a separate name for this, and then output_pk is good.

jonasnick commented on doc/musig.md:
```diff
@@ -0,0 +1,53 @@
+
+The musig API is designed to be as misuse resistant as possible.
+However, the MuSig protocol has some additional failure modes (mainly due to interactivity) that do not appear in single-signing.
+While the results can be catastrophic (e.g. leaking of the secret key), it is unfortunately not possible for the musig implementation to rule out all such failure modes.
```
done

jonasnick commented on doc/musig.md:
```diff
@@ -0,0 +1,53 @@
+## Key Aggregation and (Taproot) Tweaking
+
+Given a set of public keys, the aggregate public key is computed with `secp256k1_musig_pubkey_agg`.
+A plain tweak can be added to the resulting public key with `secp256k1_ec_pubkey_tweak_add` by setting the `tweak32` argument to the hash defined in BIP 32. Similarly, a Taproot tweak can be added with `secp256k1_xonly_pubkey_tweak_add` by setting the `tweak32` argument to the TapTweak hash defined in BIP 341.
```
done

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+
+/** Opaque data structure that caches information about public key aggregation.
+ *
+ *  Guaranteed to be 197 bytes in size. It can be safely copied/moved. No
```
I moved the "It can be safely copied/moved." to the above general section on opaque data structures; right after the sentence that the representation is platform dependent (which is the same as the doc for the other opaque data structures in the API).

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+
+/** Serialize a signer's public nonce
+ *
+ *  Returns: 1 when the nonce could be serialized, 0 otherwise
```
fixed

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,581 @@
+ *
+ *  This function differs from `secp256k1_musig_nonce_gen` by accepting a
+ *  non-repeating counter value instead of a secret random value. This requires
+ *  that the seckey argument (which is part of the keypair argument in
```
Outdated in what way?

real-or-random replied:
I like the suggestion, s/private key/secret key is more consistent.

jonasnick replied:
Replaced the sentence with the suggestion.

real-or-random commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,818 @@
+    ARG_CHECK(secp256k1_memcmp_var(&secnonce->data[0], secp256k1_musig_secnonce_magic, 4) == 0);
+    /* We make very sure that the nonce isn't invalidated by checking the values
+     * in addition to the magic. */
+    is_zero = secp256k1_is_zero_array(&secnonce->data[4], 32);
+    is_zero |= secp256k1_is_zero_array(&secnonce->data[36], 32);
```
This is what I meant. :) Yeah, it's not so much simpler, but I think it's an improvement overall.

If you want to make it one line simpler, you could do this (not sure if it's better):
```suggestion
    is_zero = secp256k1_is_zero_array(&secnonce->data[4], 2 * 32);
```

jonasnick replied:
Oh, yes! Replaced the two lines with the suggestion.

jonasnick commented on examples/musig.c:
```diff
@@ -0,0 +1,258 @@
+        pubnonces[i] = &signer[i].pubnonce;
+
+        secure_erase(seckey, sizeof(seckey));
+        secure_erase(session_secrand, sizeof(session_secrand));
```
Good catch. I removed that line. Imo explicitly erasaing secrand is more overkill than good practice.

theStack reviewed (commented):
Minor API question: are there any musig API functions where the `SECP256K1_WARN_UNUSED_RESULT` attribute should be added? (Probably for the `_{pubnonce,aggnonce,partial_sig}_parse` functions, given that external data is passed in, and the nonce generation functions?)

sipa reviewed (commented):
Code review ACK 5e55f093118c2517f068841f5414f767fb35c7fe

I have not done a detailed comparison with the BIP spec.

theStack reviewed (approved):
re-ACK 5e55f093118c2517f068841f5414f767fb35c7fe

(as per `$ git diff fdc09608036822afc1cebbe0c5b56cebf8ba508d 5e55f093118c2517f068841f5414f767fb35c7fe`)

As far as I can tell, none of the unadressed questions/comments are critical.

real-or-random commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1142 @@
+#include "../../util.h"
+
+#include "vectors.h"
+#include <inttypes.h>
```
yes, it seems that `stdint.h` suffices here

jonasnick replied:
Not needed (anymore). Removed.

real-or-random commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,817 @@
+    for (i = 0; i < 2; i++) {
+        secp256k1_gej nonce_ptj;
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &nonce_ptj, &k[i]);
+        secp256k1_ge_set_gej(&nonce_pts[i], &nonce_ptj);
```
`nonce_ptj` is guaranteed not to be infinity here. But I agree, this seems something for a follow-up.

jonasnick replied:
Why would variable time batch inversion not be acceptable here?
We could change the code to:

```C
  for (i = 0; i < 2; i++) {
      secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &nonce_ptsj[i], &k[i]);
      secp256k1_declassify(ctx, &nonce_ptsj[i], sizeof(nonce_ptsj));
      secp256k1_scalar_clear(&k[i]);
  }
  secp256k1_ge_set_all_gej_var(nonce_pts, nonce_ptsj, 2);
```

jonasnick commented:
@theStack
> are there any musig API functions where the SECP256K1_WARN_UNUSED_RESULT attribute should be added?

That's a good point. I couldn't discover a consistent pattern we're currently following (e.g, `secp256k1_ec_pubkey_parse` has the warning, `secp256k1_tagged_sha256` has the warning even though it always returns 1, and `signature_parse_der`, `signature_parse_compact` don't). I think this should be made more consistent treewide with a note to CONTRIBUTING.md.

There are different rules we could follow:
1. Just add the warning basically everywhere. Maybe that could be annoying sometimes if the user has data from a trusted source.
2. Add it to functions returning a boolean result that would be pointless to call without checking for the result (e.g., verify).
3. Add to all functions that can fail without calling a callback (e.g. they can fail because they are receiving data from untrusted sources -- at least in normal scenarios).
4. Add it to all functions you need to call before verify to make sure that verify doesn't succeed due to invalid values. However, this would be defense-in-depth because it shouldn't happen anyway.
5. Add it to all functions that operate on a seckey and produce a public output (e.g. sign) and all the functions the user needs to call before. This prevents that an invalid output (e.g., signature) accidentally leaks information about the seckey. However, users should already be protected from this because we zeroize the outputs if these functions fail.
6. Add it to functions which sound particularly scary (nonce generation).

I added it to a few more functions (following basically rules 2, 3, 4, 6), because removing the warning is backwards compatible.

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,817 @@
+        if (!secp256k1_eckey_pubkey_parse(&ges[i], &in66[33*i], 33)) {
+            return 0;
+        }
+        if (!secp256k1_ge_is_in_correct_subgroup(&ges[i])) {
+            return 0;
+        }
```
Right, added it for consistency.

jonasnick commented on include/secp256k1_musig.h:
```diff
@@ -0,0 +1,590 @@
+ *  signature for it, use `secp256k1_xonly_pubkey_tweak_add` instead.
+ *
+ *  Returns: 0 if the arguments are invalid or the resulting public key would be
+ *           invalid (only when the tweak is the negation of the corresponding
```
Correct, updated the description.

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,817 @@
+    secp256k1_nonce_function_musig(k, input_nonce, msg32, seckey, pk_ser, aggpk_ser_ptr, extra_input32);
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[0]));
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[1]));
+    VERIFY_CHECK(!secp256k1_scalar_eq(&k[0], &k[1]));
```
I don't remember. `k[0] == k[1]` doesn't lead to invalid values in contrast to the case where `k[i] == 0` (public nonce would be the unserializable point at infinity). Maybe this was added for testing that `nonce_function_musig` correctly produces two different scalars, but this is now tested explicitly.

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,817 @@
+        return 0;
+    }
+    for (i = 0; i < 2; i++) {
+        secp256k1_ge_set_gej(&aggnonce_pts[i], &aggnonce_ptsj[i]);
```
done

jonasnick commented on src/modules/musig/session_impl.h:
```diff
@@ -0,0 +1,817 @@
+    }
+
+    /* Multiply KeyAgg coefficient */
+    /* TODO Cache mu */
```
Looking at this again, I don't see an easy way to do this. Caching all coefficients would make the size of the cache dependent on the number of public keys aggregated. Right now, the cache is just a fixed size object.

Maybe caching the MuSig coefficient is overkill:
- If a signer doesn't verify partial signatures, caching all coefficients is unnecessary. They just need their own coefficient to sign. It would be possible to cache just that single coefficient, but that would require making key aggregation aware of who is going to sign with the cache later. This would be unnecessary in cases where key aggregation is done by a non-signer.
- Computing the coefficient is essentially just `to_scalar(hash(32_byte_pks_hash, serialize_33(pk))`.

jonasnick replied:
Removed the TODO

jonasnick commented on src/modules/musig/tests_impl.h:
```diff
@@ -0,0 +1,1142 @@
+    CHECK(secp256k1_schnorrsig_verify(CTX, final_sig, msg, sizeof(msg), &agg_pk) == 1);
+}
+
+static void pubnonce_summing_to_inf(secp256k1_musig_pubnonce *pubnonce) {
```
done

real-or-random reviewed (approved):
reACK 352cd3bd04ed4089cf2250b66eb73a16cb53cab3 
nit:  `musig_example` needs to be added to `.gitignore`

jonasnick commented:
> nit: musig_example needs to be added to .gitignore

fixed

real-or-random reviewed (approved):
reACK 168c92011f5ddae8d7fe28d166b68f126459548a

real-or-random commented:
We have [an open issue](https://github.com/bitcoin-core/secp256k1/issues/961#issuecomment-2397166587) for `SECP256K1_WARN_UNUSED_RESULT`, I copied @jonasnick's comment there.

theStack reviewed (approved):
re-ACK 168c92011f5ddae8d7fe28d166b68f126459548a
